### PR TITLE
Assignment7+8 linq with @BenagainB

### PR DESCRIPTION
**Pull request by mhtran23**

### CODE REVIEW

File: Assignment/Assignment/Assignment.csproj [hunk 1]
```diff
@@ -1,5 +1,20 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
-
+  <PropertyGroup>
+    <TargetFramework>net7.0</TargetFramework>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+    <LangVersion>11.0</LangVersion>
+    <nullable>enable</nullable>
+  </PropertyGroup>
```
Comment (GeorgeKoralski007) [inline Assignment/Assignment/Assignment.csproj:9]: PropertyGroup should not be here
Already in Directory.Build.props same goes for the Assignment.Tests.csproj file

**Comment by GeorgeKoralski007** [comment]
I will review

**Comment by GeorgeKoralski007** [comment]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated.✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Assignment.Tests
+{
+    [TestClass]
+    public class SampleDataTests
+    {
+
+        SampleData data = new SampleData();
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:14]:
```suggestion
        private readonly SampleData data = new();
```
IDE suggestion

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        [TestMethod]
+        public void CsvRows_NotNull()
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:17]:
```suggestion
        public void CsvRows_ReturnsNotNull()
```

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            Assert.IsNotNull(data.CsvRows);
+        }
+
+        [TestMethod]
+        public void Csv_SkipFirstRow()
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:23]:
```suggestion
        public void CsvRows_SkipFirstRow()
```
[UnitOfWork_StateUnderTest_ExpectedBehavior]

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            IEnumerable<string> row = data.CsvRows;
+            Assert.AreEqual(50, row.Count());
+        }
+
+        [TestMethod]
+        public void GetUniqueSortedStates_IsEqual()
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:30]: [UnitOfWork_StateUnderTest_ExpectedBehavior]
The name of the test could use more clearity.

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            List<string> peoplesAddress = data.GetUniqueSortedListOfStatesGivenCsvRows().ToList();
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:32]: An interesting way of doing this, but it doesn't use the System.Linq.Enumerable methods.

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            List<string> addresses = data.GetUniqueSortedListOfStatesGivenCsvRows().ToList();
+            Assert.IsTrue(peoplesAddress.SequenceEqual(addresses));
+        }
+
+        [TestMethod]
+        public void GetAggregateSortedListOfStatesUsingCsvRows_ToStringSuccess()
+        {
+            string tempString = "AL,AZ,CA,DC,FL,GA,IN,KS,LA,MD,MN,MO,MT,NC,NE,NH,NV,NY,OR,PA,SC,TN,TX,UT,VA,WA,WV";
+            string actualString = data.GetAggregateSortedListOfStatesUsingCsvRows();
+            Assert.IsTrue(tempString.Equals(actualString));
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:45]:
```suggestion
            string expected = "AL,AZ,CA,DC,FL,GA,IN,KS,LA,MD,MN,MO,MT,NC,NE,NH,NV,NY,OR,PA,SC,TN,TX,UT,VA,WA,WV";
            string actual = data.GetAggregateSortedListOfStatesUsingCsvRows();
            Assert.IsTrue(expected.Equals(actual));
```
Consider using the standard naming scheme.

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            Assert.IsNotNull(data.People.First().Address);
+        }
+
+        [TestMethod]
+        public void CreatePeople_NotNull()
+        {
+            Assert.IsNotNull(data.People);
+        }
+
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:59]: It looks like you're missing a test that leverages ISampleData.CsvRows to test ISampleData.People.

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        public void CsvRows_NotNull()
+        {
+            Assert.IsNotNull(data.CsvRows);
+        }
+
+        [TestMethod]
+        public void Csv_SkipFirstRow()
+        {
+            IEnumerable<string> row = data.CsvRows;
+            Assert.AreEqual(50, row.Count());
```
Comment (AndrewKorchemniy) [inline Assignment/Assignment.Tests/SampleDataTests.cs:26]:
```suggestion
            Assert.AreEqual<int>(50, row.Count());
```
Make sure to use the generic version

**Review by AndrewKorchemniy — COMMENTED** [review]
Sorry for the late review, I hope you can still find it helpful.

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of addresses. ✔
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔?

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
   - Be sure that `Person.Address` is also populated. ✔?
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ❌

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma-separated list of states. ✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use the `System.Linq.Enumerable.`Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.  

## Extra Credit

- Implement the homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `IAsyncSampleData`). Refactor your `SampleData` and `SampleDataAsync` classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌✔

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled  ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional)   ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version)  ❌
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

**Pull request closed without merging**